h3/src/quic.rs

```diff
@@ -11,58 +11,100 @@ use bytes::Buf;
+    fn err_code(&self) -> Option<u64>;
+}
+
+impl<'a, E: QuicError + 'a> From<E> for Box<dyn QuicError + 'a> {
+    fn from(err: E) -> Box<dyn QuicError + 'a> {
+        Box::new(err)
+    }
+}
+
 /// Trait representing a QUIC connection.
 pub trait Connection<B: Buf> {
-    /// The stream type returned when opening a unidirectional send stream.
+    /// The type produced by `poll_accept_bidi()`
+    type BidiStream: SendStream<B> + RecvStream;
+    /// The type of the sending part of `BidiStream`
     type SendStream: SendStream<B>;
-    /// The stream type returned when accepting a unidirectional receive stream.
+    /// The type produced by `poll_accept_recv()`
     type RecvStream: RecvStream;
-    /// The stream type returned when accepting or opening a bidirectional
-    /// stream.
-    ///
-    /// This type doesn't *have* to implement `BidiStream`, but if the
-    /// implementation is able to provide a cheap split, it can opt-in.
-    type BidiStream: SendStream<B> + RecvStream;
-    /// The error type that can be returned when accepting or opening a stream.
-    type Error: Into<Box<dyn std::error::Error + Send + Sync>>;
-
-    // Accepting streams
+    /// A producer of outgoing Unidirectional and Bidirectional streams.
+    type OpenStreams: OpenStreams<B>;
+    /// Error type yeilded by this trait methods
+    type Error: Into<Box<dyn QuicError>>;
 
-    /// Poll the connection for any received bidirectional streams.
+    /// Accept an incoming unidirecional stream
     ///
     /// Returning `None` implies the connection is closing or closed.
-    fn poll_accept_bidi_stream(
+    fn poll_accept_recv(
         &mut self,
         cx: &mut task::Context<'_>,
-    ) -> Poll<Result<Option<Self::BidiStream>, Self::Error>>;
+    ) -> Poll<Result<Option<Self::RecvStream>, Self::Error>>;
 
-    /// Poll the connection for any received unidirectional streams.
+    /// Accept an incoming bidirecional stream
     ///
     /// Returning `None` implies the connection is closing or closed.
-    fn poll_accept_recv_stream(
+    fn poll_accept_bidi(
         &mut self,
         cx: &mut task::Context<'_>,
-    ) -> Poll<Result<Option<Self::RecvStream>, Self::Error>>;
+    ) -> Poll<Result<Option<Self::BidiStream>, Self::Error>>;
 
-    // Opening streams
+    /// Poll the connection to create a new bidirectional stream.
+    fn poll_open_bidi(
+        &mut self,
+        cx: &mut task::Context<'_>,
+    ) -> Poll<Result<Self::BidiStream, Self::Error>>;
+
+    /// Poll the connection to create a new unidirectional stream.
+    fn poll_open_send(
+        &mut self,
+        cx: &mut task::Context<'_>,
+    ) -> Poll<Result<Self::SendStream, Self::Error>>;
+
+    /// Get an object to open outgoing streams.
+    fn opener(&self) -> Self::OpenStreams;
+
+    /// Close the connection immediately
+    fn close(&mut self, code: crate::error::Code, reason: &[u8]);
+}
+
+/// Trait for opening outgoing streams
```

Comment:
> where there's a SendRequest and Connection, right?

Yes precisely. `client::Connection` needs to be polled continuously throughout the Connection's lifetime because we need to react to incoming streams at all time: Control stream, QPACK streams and Push streams will be managed via `client::Connection` (this is what we call "driving" the connection).

Incoming streams structs such as `quinn::IncomingBi` and `quinn::IncomingUni` cannot be easily cloned, so using `Connection` directly would require a lock: 

```rust
struct client::Connection {
    inner: Arc<Mutex<impl quic::Connection>>,
}

struct client::SendRequest {
    inner: Arc<Mutex<impl quic::Connection>>,
}
```

The lock could also be managed by the quic implementation, but again this would add a superfluous lock that will be accessed all the time.

I also thought of offering the possibility to clone a `quic::Connection`, and make the new instance panic whenever a `poll_accept_*()` method is called. But this is a quite misleading API choice in my opinion.

As sending requests only require to open outgoing streams and the quinn objects for that are easily clonable, this trait appeared to be the best solution to avoid using a lock.

Still, opening a unidirectionnal stream is not very useful for sending requests, so `OpenStreams::poll_open_send()` is kind of unneeded.